### PR TITLE
Coverage bytecode filter

### PR DESCRIPTION
**Pull request by dmitrii-artuhov**
Relates to #254.
 
 This PR adds a bytecode filter class, which "hides" bytecode generated by intellij-coverage agent (e.g. when using [kover](https://github.com/Kotlin/kotlinx-kover) gradle plugin) from transformation that exist in Lincheck.
 
- [x] Hide bytecode generated by default kover settings (`hits[index] = 1`)
- [x] Hide bytecode generated by default "Run with coverage" settings in Intellij IDEA (`hits[index] += 1`)

The bytecode tracking is implemented as state machine with 5 states (`+` symbol designates the direction of an arrow):
```txt
                           INITIAL  +----------------------------------------
                       /            +                                       |
                     /                \                                     |
                   +                    +                                   |
          HITS_INIT           HITS_INIT_FIELD                               |
                    \                    /                                  |
                     +                  +                                   |
                        HITS_IN_LOCAL   -------------------------------------  
                         |         +                
                         +         |                
                      HITS_BEFORE_ASSIGN 

```

Short overview:
- The state of automata is reset when starting to instrument a new method
- `HITS_INIT` and `HITS_INIT_FIELD` states occur, when the `hits` array (which stores coverage data) is loaded onto stack. The difference between state is the bytecode that they target: the 1st one occurs when `hits` was loaded via `LDC` instruction, the 2nd when `GETSTATIC` instruction was used.
- The way intellij coverage loads `hits` using `LDC` is as follows:
  ```java
  LDC __$hits$__ : Ljava/lang/Object; ...
  CHECKCAST [Z
  ASTORE #n
  ```
  and using the `GETSTATIC`:
  ```java
  GETSTATIC CLAZZ.__$hits$__ : [Z
  IFNONNULL L0
  ...
  INVOKESTATIC com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsMask (Ljava/lang/String;)[Z
  PUTSTATIC CLAZZ.__$hits$__ : [Z
  L0
  GETSTATIC CLAZZ.__$hits$__ : [Z
  ASTORE #n
  ```
  Here `GETSTATIC` is called twice, that is why on the diagram above there is a double-sided arrow between `INITIAL` and `LOADED_ON_STACK_STATIC` states (first time `hits` is loaded on stack to check if it is null, then it is discarded, and lastly it is put on stack once again).
- `HITS_IN_LOCAL` occurs when the `hits` array is already saved to local variables of the method and the automata knows its index.
- When the `ALOAD #n` instruction occurs automata goes into `HITS_BEFORE_ASSIGN` state, then it searches for bytecode of `hits[index] = 1` or/and `hits[index] += 1`. When the assignment completed the state goes back to `HITS_IN_LOCAL`.
- When the method is fully transformed, the automata is put to `INITIAL` state again.

### CODE REVIEW

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 1]
```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
...
+
+        override fun visitEnd() {
+            state = State.INITIAL
+            localVariableIndex = -1
+
+            super.visitEnd()
+        }
+    }
+
+    private class CoroutineCancellabilitySupportMethodTransformer(
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:256]: Is this related to this PR?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 2]
```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
...
+                        initialAdapter.visitFieldInsn(opcode, owner, name, desc)
+                        return
+                    }
+                }
+            }
+            super.visitFieldInsn(opcode, owner, name, desc)
+        }
+
+        override fun visitJumpInsn(opcode: Int, label: Label?) {
+            when (opcode) {
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:184]: `if (opcode == IFNONNULL)`

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 3]
```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
...
+     * @param initialAdapter initial adapter to delegate work to when coverage-related bytecode encountered.
+     * @param nextAdapter adapter to delegate calls to when bytecode is not related to coverage.
+     */
+    private class CoverageBytecodeFilter(
+        private val initialAdapter: GeneratorAdapter,
+        nextAdapter: GeneratorAdapter
+    ) : MethodVisitor(ASM_API, nextAdapter) {
+        enum class State {
+            INITIAL,
+            LOADED_HITS_ON_STACK,
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:148]: I would suggest some renaming, what do you think?
```
LOADED_HITS_ON_STACK -> INIT
LOADED_HITS_ON_STACK_STATIC -> INIT_FIELD
LOADED_HITS_IN_LOCALS -> IN_LOCAL
PREPARE_ASSIGN_TO_HITS -> ON_STACK
```

**Comment by zuevmaxim** [comment]
Did you consider implementing similar filter for JaCoCo?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 4]
```diff
@@ -26,8 +26,8 @@ internal class LincheckClassVisitor(
     private val ideaPluginEnabled = ideaPluginEnabled()
     private var classVersion = 0
 
-    private lateinit var fileName: String
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:29]: Are these changes needed?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt (hunk 3 above)
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:148]: Should we then start every enum variant with `HITS` prefix?

```
HITS_INIT
HITS_INIT_FIELD
HITS_IN_LOCAL
HITS_ON_STACK
```

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 5]
```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
...
+    ) : MethodVisitor(ASM_API, nextAdapter) {
+        enum class State {
+            INITIAL,
+            LOADED_HITS_ON_STACK,
+            LOADED_HITS_ON_STACK_STATIC,
+            LOADED_HITS_IN_LOCALS,
+            PREPARE_ASSIGN_TO_HITS,
+        }
+        private val COVERAGE_HITS_NAME = "__\$hits\$__"
+        private val COVERAGE_GET_HITS_MASK_METHOD_SIGNATURE = "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsMask(Ljava/lang/String;)[Z"
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:154]: For the `hits[i]++` version, here will be 
```
com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHits (Ljava/lang/String;)[I
```
You can check all options here https://github.com/JetBrains/intellij-coverage/tree/master/test-kotlin/resources/bytecode/simple/branches

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt (hunk 1 above)
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:256]: `CoroutineCancellabilitySupportMethodTransformer` was moved into `transformers/CoroutineSupportTransformers.kt` file, please remove it from here.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 6]
```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
...
     }
 
 }
+    /**
+     * Filters out chunks of bytecode generated by coverage.
+     *
+     * @param initialAdapter initial adapter to delegate work to when coverage-related bytecode encountered.
+     * @param nextAdapter adapter to delegate calls to when bytecode is not related to coverage.
+     */
+    private class CoverageBytecodeFilter(
```
Comment (eupp) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:142]: Please move `CoverageBytecodeFilter` into a separate file in the `transformers` sub-package, similarly as it is done for other byte-code transformers.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt [hunk 7]
```diff
@@ -121,13 +122,159 @@ internal class LincheckClassVisitor(
             sv.analyzer = aa
             aa
         }
+        mv = CoverageBytecodeFilter(
```
Comment (eupp) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckClassVisitor.kt:125]: Please add a comment warning that `CoverageBytecodeFilter` should be placed after `SharedMemoryAccessTransformer` 

(Perhaps the relative order to some other transformers is also important? If so, please also mention this.)

**Review by eupp — CHANGES_REQUESTED** [review]

**Comment by dmitrii-artuhov** [comment]
> Did you consider implementing similar filter for JaCoCo?

For this PR I did not consider that. I can add filtering for JaCoCo instrumentation in another PR, I guess. @eupp what do you think?

**Comment by eupp** [comment]
@zuevmaxim what is a typical workflow of kover? 
Does it support JaCoCo? What coverage method is used by default: intellij-coverage or JaCoCo? How often JaCoCo is used by the developers as a backend for kover?

If JaCoCo is not used very often, I would propose to postpone implementing Lincheck integration for it.

**Comment by dmitrii-artuhov** [comment]
> Does it [kover] support JaCoCo?
Yes

> What coverage method is used by default: intellij-coverage or JaCoCo?
intellij-coverage

Not sure about the spread among users who utilize different backends.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt [hunk 8]
```diff
@@ -327,8 +332,22 @@ internal object LincheckClassFileTransformer : ClassFileTransformer {
         val reader = ClassReader(classBytes)
         val writer = SafeClassWriter(reader, loader, ClassWriter.COMPUTE_FRAMES)
         try {
-            reader.accept(LincheckClassVisitor(instrumentationMode, writer), ClassReader.SKIP_FRAMES)
-            writer.toByteArray()
+            //println("Initial class bytecode for '$className'")
```
[the remark this answers is not in the document]
Comment (dmitrii-artuhov) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt]: I added some comments to this code, that say that the code might be used for debugging

**Comment by zuevmaxim** [comment]
@eupp Within kover workflow, JaCoCo is not used often

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/transformers/CoverageBytecodeFilter.kt [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        HITS_BEFORE_ASSIGN,
+    }
+    private val COVERAGE_HITS_NAME = "__\$hits\$__"
+    /** @see <a href="https://github.com/JetBrains/intellij-coverage/blob/master/test-kotlin/resources/bytecode/simple/branches/branch_condy.txt">Hits masks static methods</a> */
+    private val COVERAGE_HITS_METHOD_SIGNATURES = listOf(
+        "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsMask(Ljava/lang/String;)[Z",
+        "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsMaskCached(Ljava/lang/String;)[Z",
+        "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHits(Ljava/lang/String;)[I",
+        "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsCached(Ljava/lang/String;)[I",
+        "com/intellij/rt/coverage/instrumentation/CoverageRuntime.getHitsMaskCached(Ljava/lang/String;)[Z",
```
Comment (zuevmaxim) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/transformers/CoverageBytecodeFilter.kt:45]: This is a duplicate

**Review by zuevmaxim — APPROVED** [review]

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (eupp) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt]: We need a more rigorous and systematic approach for byte-code transformation testing and debugging. 
Let's remove this code from this PR, and revisit the problem later under #330

**Review by eupp — APPROVED** [review]

**Comment by eupp** [comment]
Please rebase on develop and re-open under a Lincheck repository branch to enable CI checks. Then assign @ndkoval as a reviewer to the new PR.
